Add options: safe_limit, show_query and expand_to_paragraph

optio opened the pull request:
Define a minimum limit to avoid accidental big and slow queries.
If query is a SELECT and doesn't contain a LIMIT and safe_limit is set, then add a LIMIT
option= safe_limit

&

Show above the results the executed query, including a timestamp and duration
option= show_query

optio commented:
Maybe the safe_limit should be default False instead of 1000?

mtxr commented:
I'm testing it, just give me some time and I'll approve.

Thanks, great idea!

optio commented:
Found an issue when the selected query without a limit is ending with a ';' , will commit fix to this request

optio commented:
Done , though can you suggest how to refactore this to more elegant code ?

Btw, used '-15' as that seems a safe 'bet' to position any potential 'limit' or ';'

            if query.lower().startswith('select'):
                if " limit " not in query.lower()[-15:]:
                    if self.safe_limit:
                        if ";" in query.lower()[-5:]:
                            seperatorIndex = query.rfind(';')
                            query = query[:seperatorIndex] + query[seperatorIndex+1:]
                        query += " LIMIT {0}".format(self.safe_limit)

optio commented:
I combined this request for the options "safe_limit" and "show_query" (and closed pullrequest #56)

mtxr commented:
Continuing the comments from code review [Here]:(https://github.com/mtxr/SQLTools/pull/55/files#r92516358)

But I was thinking here if, instead of adding the limit, we could allow the user to add anything to any query, as a prefix and suffix. Like this:


On settings:
```
...
"query_prefix": "BEGIN;"
"query_suffix": "LIMIT 10; ROLLBACK;"
...
```

Then, it would be concatenated to the query. It's an idea, any suggestions?

optio commented:
I'll try another method to parse the query, to detect whether a real 'limit' reserved word was added.

Regarding the query_prefix and query_suffix, that could be a great feature too, though think it's unrelated to this "safe_limit' feature. Since the safe_limit would only apply if the user didn't provide a limit, not upon every query.

mtxr commented:
@optio hey! 
I think I found a way to help you!

https://pypi.python.org/pypi/sqlparse, we already use this lib, look at the parsing section:
```
>>> sql = 'select * from someschema.mytable where id = 1'
>>> res = sqlparse.parse(sql)
>>> res
(<Statement 'select...' at 0x9ad08ec>,)
>>> stmt = res[0]
>>> str(stmt)  # converting it back to unicode
'select * from someschema.mytable where id = 1'
>>> # This is how the internal representation looks like:
>>> stmt.tokens
(<DML 'select' at 0x9b63c34>,
 <Whitespace ' ' at 0x9b63e8c>,
 <Operator '*' at 0x9b63e64>,
 <Whitespace ' ' at 0x9b63c5c>,
 <Keyword 'from' at 0x9b63c84>,
 <Whitespace ' ' at 0x9b63cd4>,
 <Identifier 'somes...' at 0x9b5c62c>,
 <Whitespace ' ' at 0x9b63f04>,
 <Where 'where ...' at 0x9b5caac>)
```

optio commented:
I tried using parsesql , and that is indeed the best way, though I'm still not able to parse a query.
See below example:
The 1st rawQuery parses fine, though the 2nd query doesn't parse. The 3rd does parse, it's the same as the 2nd, though with the quotes in the query removed (thus the 3rd is an invalid sql query)


import pprint
import sqlparse

rawQueries = [
    "select * from table_name left join sometable on this.id = table_name.testid where property ilike '% limit %' limit 10; select * from table_name limit 2",
    "select * from table_name where name ilike '% limit %' ;   select * from table_name",
    "select * from table_name where name ilike % limit % ;   select * from table_name",
]

for rawQuery in rawQueries:
    print('** New RAW query **')
    print(rawQuery)
    # rawQuery = str(sqlparse.format(rawQuery, reindent=False, keyword_case='upper'))
    # for query in sqlparse.split(rawQuery.replace("'", "")):
    for query in sqlparse.split(rawQuery):
        print('--New query --')
        for parse in sqlparse.parse(query):
            for token in parse.tokens:
                if token.ttype in sqlparse.tokens.Keyword and token.value == 'limit':
                    print ("Found LIMIT in " + query)
                    pprint.pprint(token)
                    pprint.pprint(token.ttype)
                    pprint.pprint(token.value)
                if token.ttype in sqlparse.tokens.Punctuation and token.value == ';':
                    print ("Found PUNTUATION in " + query)
                    pprint.pprint(token)
                    pprint.pprint(token.ttype)
                    pprint.pprint(token.value)

mtxr commented:
I'll try some stuff during the weekend too, ok?
Let's make it work :D

optio commented:
Please check my latest attempt. Uses sqlparse for 'select' and 'no limit' detection.
Though:
- The select detection sometimes fails so there's still a plain 'string starts with select' fallback
- All single quoates ' are replaces into double quotes `sqlparse.parse(query.strip().replace("'", "\""))` to make above described scenatiio's work.

mtxr commented:
Seems to be working! I'll be testing it tonight and approving by tomorrow.

Nice job!

optio commented:
Added a new option : 'expand_to_paragraph':

If user has no text selected and extend_to_paragraph is enabled, then get the current paragraph instead of only the current line.

r-stein commented on SQLTools.py:
```diff
@@ -132,7 +133,10 @@ def getSelection():
         for region in View().sel():
             if region.empty():
-                text.append(View().substr(View().line(region)))
+                if not settings.get('expand_to_paragraph', False):
```
If you use a setting you should also add it to the settings file.

optio replied:
it's added in the next commit, https://github.com/optio/SQLTools/commit/146141e827e232f39bc42928e781ca831bf25f05 .

optio commented:
@mtxr FYI : https://forum.sublimetext.com/t/select-current-block-of-text-solved/25118/9

optio commented:
Bump version for package control?

optio commented:
Also, http://code.mteixeira.me/SQLTools/ RE Settings needs a small update to mention the new options

optio commented:
Added 2 new pull request for readme/website & version bump
